quipucords/api/credential/view.py

```diff
@@ -0,0 +1,173 @@
+    AuthTokenSerializerV2,
+    CredentialSerializerV2,
+    SshCredentialSerializerV2,
+    UsernamePasswordSerializerV2,
+)
+from api.credential.view_v1 import CredentialFilter
+from constants import DataSources
+
+
+@api_view(["post"])
+def credential_bulk_delete(request):
+    """
+    Bulk delete credentials.
+
+    Response payload contains IDs of credentials deleted, skipped, and not found.
+    Example response:
+
+        {
+            "message": \
+                "Deleted 3 credentials. "\
+                "Could not find 0 credentials. "\
+                "Failed to delete 2 credentials.",
+            "deleted": [1, 2, 3],
+            "missing": [],
+            "skipped": [
+                {"credential": 6, "sources": [1]},
+                {"credential": 7, "sources": [2, 3]},
+            ],
+        }
+
+    input:      "ids" : List of ids to delete, or string ALL_IDS_MAGIC_STRING
+    returns:    200 OK - upon successfully deleting any credentials.
+                400 Bad Request - ids list is missing or empty.
+    """
+    ids = set_of_ids_or_all_str(request.data.get("ids"))
+
+    with transaction.atomic():
+        creds = Credential.objects.all()
+        if ids != ALL_IDS_MAGIC_STRING:
+            creds = creds.filter(id__in=ids)
+        credential_ids_requested = ids if isinstance(ids, set) else set()
+        credential_ids_found = set(creds.values_list("id", flat=True))
+        credential_ids_with_sources = (
+            creds.exclude(sources=None)
+            .prefetch_related("sources")
+            .values_list("id", "sources")
+            .order_by("id")  # later groupby needs sorted input
+        )
+        creds.filter(sources=None).delete()
+
+    credential_ids_missing = credential_ids_requested - credential_ids_found
+    credential_ids_skipped = []
+
+    for credential_id, grouper in groupby(
+        credential_ids_with_sources, key=lambda c: c[0]
+    ):
+        credential_ids_skipped.append(
+            {
+                "credential": credential_id,
+                "sources": [g[1] for g in grouper],
+            }
+        )
+    credential_ids_deleted = credential_ids_found - set(
+        c["credential"] for c in credential_ids_skipped
+    )
+
+    message = _(
+        "Deleted {count_deleted} credentials. "
+        "Could not find {count_missing} credentials. "
+        "Failed to delete {count_failed} credentials."
+    ).format(
+        count_deleted=len(credential_ids_deleted),
+        count_missing=len(credential_ids_missing),
+        count_failed=len(credential_ids_skipped),
+    )
+    response_data = {
+        "message": message,
+        "deleted": credential_ids_deleted,
+        "missing": credential_ids_missing,
+        "skipped": credential_ids_skipped,
+    }
+    return Response(data=response_data, status=status.HTTP_200_OK)
+
+
+class CredentialViewSetV2(ModelViewSet):
+    """A view set for the Credential model."""
+
+    queryset = Credential.objects.all()
```

Comment:
I don't think I understand. Would you elaborate what "N+1 problem" are you talking about here?

Reply:
for future reference, for each loaded credential django will also query for related sources when we use our serializers. This is problematic for the "list" view - with the default page size  = 100, we would hit the db 101 times instead of just 1 or 2 (given this is m2m I think django will make 2 queries instead of a single one, but I'm not a 100% sure on that).

Reply:
I amended the commit that (re)introduced this code, and I added `prefetch_related` to `queryset`. I read about the `Prefetch` object, and I don't think we have a case for it here since we do not have any additional filters to apply for the related Sources, but I'm happy to take a second look if you know of a use case that I don't.